src/Rules/Test.hs

```diff
@@ -28,45 +28,116 @@ testRules = do
-            , "-e", "ghc_with_vanilla=1" -- TODO: do we always build vanilla?
-            , "-e", "ghc_with_dynamic=0" -- TODO: support dynamic
-            , "-e", "ghc_with_profiling=0" -- TODO: support profiling
-            , "-e", "ghc_with_interpreter=" ++ show ghcWithInterpreterInt
-            , "-e", "ghc_unregisterised=" ++ show ghcUnregisterisedInt
-            , "-e", "ghc_with_threaded_rts=0" -- TODO: support threaded
-            , "-e", "ghc_with_dynamic_rts=0" -- TODO: support dynamic
-            , "-e", "ghc_dynamic_by_default=False" -- TODO: support dynamic
-            , "-e", "ghc_dynamic=0" -- TODO: support dynamic
-            , "-e", "ghc_with_llvm=0" -- TODO: support LLVM
-            , "-e", "in_tree_compiler=True" -- TODO: when is it equal to False?
-            , "-e", "clean_only=False" -- TODO: do we need to support True?
-            , "--configfile=testsuite/config/ghc"
-            , "--config", "compiler=" ++ show (top -/- compiler)
-            , "--config", "ghc_pkg="  ++ show (top -/- ghcPkg)
-            , "--config", "haddock="  ++ show (top -/- haddock)
-            , "--summary-file", "testsuite_summary.txt"
-            , "--threads=" ++ show threads
+        haddock  <- builderPath $ Haddock BuildPackage
+        hp2ps    <- builderPath $ Hp2Ps
+        hpc      <- builderPath $ Hpc
+
+        let minGhcVersion v = (v <=) <$> ghcCanonVersion
+
+        let ifMinGhcVer v opt = do ver <- ghcCanonVersion
+                                   if v <= v then pure opt
+                                             else pure ""
+
+        -- Prepare extra flags to send to the Haskell compiler.
+        -- TODO: read extra argument for test from command line, like `-fvectorize`.
+        let ghcExtraFlags = if unregisterised               -- Value EXTRA_HC_OPTS should be handled.
+                               then "-optc-fno-builtin"
+                               else ""
+
+        -- Take flags to send to the Haskell compiler from test.mk.
+        -- See: https://github.com/ghc/ghc/blob/cf2c029ccdb967441c85ffb66073974fbdb20c20/testsuite/mk/test.mk#L37-L55
+        ghcFlags <- sequence
+            [ pure "-dcore-lint -dcmm-lint -no-user-package-db -rtsopts"
+            , pure ghcExtraFlags
+            , ifMinGhcVer "711" "-fno-warn-missed-specialisations"
+            , ifMinGhcVer "711" "-fshow-warning-groups"
+            , ifMinGhcVer "801" "-fdiagnostics-color=never"
+            , ifMinGhcVer "801" "-fno-diagnostics-show-caret"
+            , pure "-dno-debug-output"
             ]
 
-            -- , "--config", "hp2ps="    ++ quote ("hp2ps")
-            -- , "--config", "hpc="      ++ quote ("hpc")
-            -- , "--config", "gs=$(call quote_path,$(GS))"
-            -- , "--config", "timeout_prog=$(call quote_path,$(TIMEOUT_PROGRAM))"
+        -- See: https://github.com/ghc/ghc/blob/master/testsuite/mk/test.mk#L291
+        let timeout_prog = "testsuite/timeout/install-inplace/bin/timeout"
```

Comment:
Does this therefore rely on having the timeout prog make rules executed before running hadrian's `test` rule? I do have a rule or two to handle the timeout program in my branch. I don't properly implement the Windows logic yet, I always pick the python script, but that's a start.

Reply:
If the `timeout_prog` is not set, I get a `[Error 13] Permission denied` in python script (inside the `runCmd` method).

I'm working with Bash on Windows, it's a Ubuntu 16.04 environment.

Reply:
We have the following code in `testlib.py`:

```
        r = subprocess.Popen([timeout_prog, timeout, cmd],
                             stdin=stdin_file,
                             stdout=subprocess.PIPE,
                             stderr=hStdErr,
                             env=ghc_env)
```

Reply:
Right, so this indeed requires that you build it through the make build system first. I have a bit of code that works on Linux for wrapping the python version of the program in a shell script and using that for the tests. It's a bit ugly but it does the trick. You'd need to replicate the Windows logic to make this work on your Windows machine, but as you can see [here](https://github.com/ghc/ghc/blob/master/testsuite/timeout/Makefile#L36) it's not scary enough to prevent us from doing it all in hadrian, I think. Here's the code that I have for the non-Windows case (that is, the part that replicates [this bit](https://github.com/ghc/ghc/blob/master/testsuite/timeout/Makefile#L23) of the Makefile).

``` haskell
-- FIXME: add ".exe" extension on windows
timeoutProgPath :: FilePath
timeoutProgPath = "test" -/- "bin" -/- "timeout"

timeoutPyPath :: FilePath
timeoutPyPath = "test" -/- "bin" -/- "timeout.py"

-- and then in `testRules`...
    root <- buildRootRules

    root -/- timeoutPyPath ~> do -- FIXME: run only when not on Windows
      copyFile "testsuite/timeout/timeout.py" (root -/- timeoutPyPath)

    root -/- timeoutProgPath ~> do
      need [ root -/- timeoutPyPath ]
      let script = unlines
            [ "#!/usr/bin/env sh"
            , "exec python3 $0.py \"$@\""
            ]
      liftIO $ do
        writeFile (root -/- timeoutProgPath) script
        cmd "chmod" [ "+x", root -/- timeoutProgPath ]
```

Reply:
If we can get ride of current python's test driver, we could do the `timeout` directly in Haskell.

Reply:
I'm afraid the python test driver is not going anytime soon, and we might want to aim for a more reasonable goal of just replicating that tiny Makefile's logic in hadrian, so that we can at least properly build/wrap the appropriate timeout program but also `need` it in any rule that will run tests. Really just replicating what's done in the makefile I linked to but in hadrian, so that we don't introduce any (implicit) dependency on them make build system for running the test rules.